numba/rewrites.py

```diff
@@ -0,0 +1,42 @@
+from __future__ import print_function, division, absolute_import
+
+
+class Rewrite(object):
+    def match(self, block, typemap, calltypes):
+        return False
+
+    def apply(self):
+        raise NotImplementedError("Abstract Rewrite.apply() called!")
+
+
+class RewriteRegistry(object):
+    def __init__(self):
+        self.rewrites = []
+
+    def register(self, rewrite_cls):
+        assert issubclass(rewrite_cls, Rewrite)
+        self.rewrites.append(rewrite_cls)
+        return rewrite_cls
+
+    def apply(self, pipeline, blocks):
+        old_blocks = blocks.copy()
+        for rewrite_cls in self.rewrites:
+            # Exhaustively apply a rewrite until it stops matching.
+            rewrite = rewrite_cls(pipeline)
+            work_list = list(blocks.items())
+            while work_list:
+                key, block = work_list.pop()
+                matches = rewrite.match(block, pipeline.typemap,
```

Comment:
This division between match/apply is an artifact of how I think about rewriting languages.  When using a rewriting DSL, you commonly have a pattern match half, and a replacement term construction half.  In the rewriting DSL, there is a namespace that gets built by pattern matching, and this namespace gets passed to the constructor.  Here, internal state replaces that namespace, but the pattern still "works".  Keeping these separate may also give us more flexibility in the future when composing rewrites.